src/ScopeProvider/scope.ts

```diff
@@ -0,0 +1,115 @@
+import { atom, type Atom } from 'jotai';
+import { type AnyAtom, type AnyWritableAtom } from './types';
+
+export type Scope = {
+  getAtom: <T extends AnyAtom>(originalAtom: T, fromScoped?: boolean) => T;
+};
+
+export function createScope(
+  atoms: Iterable<AnyAtom>,
+  parentScope: Scope | undefined,
+): Scope {
+  const explicit = new WeakMap<AnyAtom, AnyAtom>();
+  const implicit = new WeakMap<AnyAtom, AnyAtom>();
+  const inherited = new WeakMap<AnyAtom, AnyAtom>();
+  const unscoped = new WeakMap<AnyAtom, AnyAtom>();
+
+  // populate explicit scope map
+  for (const anAtom of atoms) {
+    const scopedCopy = copyAtom(anAtom, true);
+    explicit.set(anAtom, scopedCopy);
+  }
+
+  /**
+   * Creates a scoped atom from the original atom.
+   * All dependencies of the original atom will be implicitly scoped.
+   * All scoped atoms will be inherited scoped in nested scopes.
+   * @param anAtom
+   * @param isScoped - the atom is a dependency of an explicitly scoped atom
+   * @returns
+   */
+  function copyAtom<T>(anAtom: Atom<T>, isScoped = false) {
+    // avoid reading `init` to preserve lazy initialization
+    const scopedAtom: Atom<T> = Object.create(
+      Object.getPrototypeOf(anAtom),
+      Object.getOwnPropertyDescriptors(anAtom),
+    );
+
+    if ('read' in scopedAtom) {
+      // inherited atoms should preserve their value
+      if (scopedAtom.read === defaultRead) {
+        const self = isScoped ? scopedAtom : anAtom;
+        scopedAtom.read = defaultRead.bind(self) as Atom<T>['read'];
+      } else {
+        scopedAtom.read = (get, opts) => {
+          return anAtom.read((a) => {
+            return get(getAtom(a, isScoped));
+          }, opts);
+        };
+      }
+    }
+
+    if ('write' in scopedAtom) {
+      if (scopedAtom.write === defaultWrite) {
+        const self = isScoped ? scopedAtom : anAtom;
+        scopedAtom.write = defaultWrite.bind(self);
+      } else {
+        (scopedAtom as AnyWritableAtom).write = (get, set, ...args) => {
+          return (anAtom as AnyWritableAtom).write(
+            (a) => get(getAtom(a, isScoped)),
+            (a, ...v) => set(getAtom(a, isScoped), ...v),
+            ...args,
+          );
+        };
+      }
+    }
+
+    // eslint-disable-next-line camelcase
+    scopedAtom.unstable_is = (a: AnyAtom) => {
```

Comment:
FIXME: there is some infinite loop with unstable_is and inherited atoms I still need to resolve.